// ride-css/index.js
// Ride css. Released under MIT licence.
// https://github.com/OctoD/ride-css/
var Fibonacci = require('./fibonacci')
  , Gradients = require('./gradients')
  , Grid = require('./grid')
  , Typography = require('./typography');

// exporting a fibonacci sequence object
module.exports.fibonacci = function (sequenceStopsAt) {
  var fibonacciInstance = new Fibonacci(sequenceStopsAt);

  return fibonacciInstance.getSequence();
}

// exposing the gradients 
module.exports.gradients = function (config) {

}

// exporting the grid Object
module.exports.grid = function (width, columns, gutter, padding) {
  var gridInstance = new Grid(parseFloat(width) || 100, parseFloat(columns) || 12, gutter || [0], padding || [0]);
  
  return {
    cols: gridInstance.getColumns(),
    floats: gridInstance.getFloats(),
    instance: gridInstance,
    gutter: gridInstance.gutter,
    padding: gridInstance.padding,
    subtractGutter: gridInstance.calcGutter(),
    subtractPadding: gridInstance.calcGutter()
  }
}

// exporting the typography Object
module.exports.typography = function (fontSize, lineHeight) {
  var typographyInstance;

  fontSize = (Object.prototype.toString.call(fontSize) === '[object Array]') ? fontSize : [fontSize];

  typographyInstance = new Typography(fontSize, lineHeight);

  return {
    headings: typographyInstance.getHeadings(),
    instance: typographyInstance,
    paragraph: typographyInstance.getParagraph()
  }
}
